File: backend/app/schemas/quick_analyze.py
"""Pydantic schemas for the Quick Analyze workflow."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field

QuickAnalyzeStage = Literal[
    "created",
    "uploaded",
    "profiled",
    "planned",
    "awaiting_confirmation",
    "running",
    "complete",
    "failed",
]

AnalysisIntentChoice = Literal["association", "exploratory"]


class QuickAnalyzeSession(BaseModel):
    session_id: str
    stage: QuickAnalyzeStage
    project_id: str | None = None
    dataset_id: str | None = None
    plan_id: str | None = None
    analysis_id: str | None = None
    research_question: str | None = None
    analysis_intent: AnalysisIntentChoice = "association"
    recommendation: "RecommendationCard | None" = None
    summary: "BeginnerSummary | None" = None
    progress_message: str = ""
    error_message: str | None = None
    created_at: datetime
    updated_at: datetime


class RecommendationCard(BaseModel):
    """One-screen summary of the recommended analysis for user confirmation."""
    outcome_variable: str
    main_variable: str
    control_variables: list[str] = Field(default_factory=list)
    detected_structure: str
    recommended_model: str
    recommended_model_type: str
    transformation_suggestions: list[str] = Field(default_factory=list)
    why_reasons: list[str] = Field(default_factory=list)
    warnings: list[str] = Field(default_factory=list)
    is_exploratory: bool = False
    needs_user_input: str | None = None


class ConfirmationRequest(BaseModel):
    dependent_variable: str
    primary_independent_variable: str
    control_variables: list[str] = Field(default_factory=list)
    entity_column: str | None = None
    time_column: str | None = None
    model_type: str
    apply_log_transform_to: list[str] = Field(default_factory=list)
    analysis_intent: AnalysisIntentChoice = "association"


class DiagnosticsStatusCard(BaseModel):
    data_quality: Literal["Good", "Needs review"]
    model_fit: Literal["Available", "Limited"]
    multicollinearity: Literal["Low concern", "Moderate concern", "High concern"]
    heteroskedasticity: Literal["Not detected", "Detected — robust standard errors recommended"]
    panel_structure: Literal["Detected", "Not detected"]
    causal_interpretation: Literal["Association only"]


class BeginnerSummary(BaseModel):
    """Plain-language summary of analysis results for non-technical users."""
    headline: str
    dataset_description: str
    model_used: str
    main_finding: str
    is_significant: bool
    significance_threshold: float = 0.05
    causal_warning: str
    key_warnings: list[str] = Field(default_factory=list)
    diagnostics_status: DiagnosticsStatusCard
    next_actions: list[str] = Field(default_factory=list)


class QuickAnalyzeUploadResponse(BaseModel):
    session_id: str
    dataset_id: str
    project_id: str
    filename: str
    n_rows: int
    n_columns: int
    stage: QuickAnalyzeStage


class QuickAnalyzePlanResponse(BaseModel):
    session_id: str
    stage: QuickAnalyzeStage
    recommendation: RecommendationCard
    progress_message: str


class QuickAnalyzeRunResponse(BaseModel):
    session_id: str
    stage: QuickAnalyzeStage
    analysis_id: str
    summary: BeginnerSummary
    progress_message: str


class QuickAnalyzeSessionDetail(BaseModel):
    session: QuickAnalyzeSession
    workspace_url: str | None = None

File: frontend/types/quick_analyze.ts
export type QuickAnalyzeStage =
  | "created"
  | "uploaded"
  | "profiled"
  | "planned"
  | "awaiting_confirmation"
  | "running"
  | "complete"
  | "failed";

export type AnalysisIntentChoice = "association" | "exploratory";

export interface RecommendationCard {
  outcome_variable: string;
  main_variable: string;
  control_variables: string[];
  entity_column: string | null;
  time_column: string | null;
  detected_structure: string;
  recommended_model: string;
  recommended_model_type: string;
  transformation_suggestions: string[];
  why_reasons: string[];
  warnings: string[];
  is_exploratory: boolean;
  needs_user_input: string | null;
}

export interface DiagnosticsStatusCard {
  data_quality: "Good" | "Needs review";
  model_fit: "Available" | "Limited";
  multicollinearity: "Low concern" | "Moderate concern" | "High concern";
  heteroskedasticity:
    | "Not detected"
    | "Detected — robust standard errors recommended";
  panel_structure: "Detected" | "Not detected";
  causal_interpretation: "Association only";
}

export interface BeginnerSummary {
  headline: string;
  dataset_description: string;
  model_used: string;
  main_finding: string;
  is_significant: boolean;
  significance_threshold: number;
  causal_warning: string;
  key_warnings: string[];
  diagnostics_status: DiagnosticsStatusCard;
  next_actions: string[];
}

export interface QuickAnalyzeSession {
  session_id: string;
  stage: QuickAnalyzeStage;
  project_id: string | null;
  dataset_id: string | null;
  plan_id: string | null;
  analysis_id: string | null;
  research_question: string | null;
  analysis_intent: AnalysisIntentChoice;
  recommendation: RecommendationCard | null;
  summary: BeginnerSummary | null;
  progress_message: string;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface QuickAnalyzeUploadResponse {
  session_id: string;
  dataset_id: string;
  project_id: string;
  filename: string;
  n_rows: number;
  n_columns: number;
  stage: QuickAnalyzeStage;
}

export interface QuickAnalyzePlanResponse {
  session_id: string;
  stage: QuickAnalyzeStage;
  recommendation: RecommendationCard;
  progress_message: string;
}

export interface ConfirmationRequest {
  dependent_variable: string;
  primary_independent_variable: string;
  control_variables: string[];
  entity_column: string | null;
  time_column: string | null;
  model_type: string;
  apply_log_transform_to: string[];
  analysis_intent: AnalysisIntentChoice;
}

export interface QuickAnalyzeRunResponse {
  session_id: string;
  stage: QuickAnalyzeStage;
  analysis_id: string;
  summary: BeginnerSummary;
  progress_message: string;
}

export interface QuickAnalyzeSessionDetail {
  session: QuickAnalyzeSession;
  workspace_url: string | null;
}
